### PR TITLE
Mediation adapters update

### PR DESCRIPTION
**Pull request by LTPhantom**
## Description

Adds `MediationExtras` class to include parameters when using mediation through the implementation of `FlutterMediationExtras` in Android and `FlutterMediationExtras` in iOS. It also deprecates the use of the Network Extras Provider for both platforms.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/googleads/googleads-mobile-flutter/issues
[Contributor Guide]: https://github.com/googleads/googleads-mobile-flutter/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/google_mobile_ads/android/src/main/java/io/flutter/plugins/googlemobileads/FlutterMediationExtras.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+// limitations under the License.
+
+package io.flutter.plugins.googlemobileads;
+
+import android.os.Bundle;
+import android.util.Pair;
+import com.google.android.gms.ads.mediation.MediationExtrasReceiver;
+import java.util.Map;
+
+public interface FlutterMediationExtras {
```
Comment (jjliu15) [inline packages/google_mobile_ads/android/src/main/java/io/flutter/plugins/googlemobileads/FlutterMediationExtras.java:22]: Can you add a comment describing how this class is expected to be used? E.g. mediation plugins will provide implementations

File: packages/google_mobile_ads/ios/Classes/FLTMediationExtras.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import <Foundation/Foundation.h>
+#import <GoogleMobileAds/GADAdNetworkExtras.h>
+
+@protocol FlutterMediationExtras
```
Comment (jjliu15) [inline packages/google_mobile_ads/ios/Classes/FLTMediationExtras.h:18]: Same here, could you add some comments outlining how this is expected to be used?

File: packages/google_mobile_ads/lib/src/ad_containers.dart [hunk 3]
```diff
@@ -190,15 +192,15 @@ class LoadAdError extends AdError {
...
-    this.mediationExtrasIdentifier,
-    this.extras,
-  });
+  const AdRequest(
+      {this.keywords,
+      this.contentUrl,
+      this.neighboringContentUrls,
+      this.nonPersonalizedAds,
+      this.httpTimeoutMillis,
+      this.mediationExtrasIdentifier,
```
Comment (jjliu15) [inline packages/google_mobile_ads/lib/src/ad_containers.dart:201]: Can you mark this as deprecated?

File: packages/google_mobile_ads/lib/src/ad_containers.dart [hunk 4]
```diff
@@ -233,6 +235,10 @@ class AdRequest {
   /// Extras to pass to the AdMob adapter.
   final Map<String, String>? extras;
 
+  /// Extras to pass to the Mediations Adapter linked to the instance of the
```
Comment (jjliu15) [inline packages/google_mobile_ads/lib/src/ad_containers.dart:238]: Suggestion: "Extra parameters to pass to specific ad adapters."

File: packages/google_mobile_ads/lib/src/ad_instance_manager.dart [hunk 5]
```diff
@@ -1053,7 +1065,13 @@ class AdMessageCodec extends StandardMessageCodec {
           mediationExtrasIdentifier: readValueOfType(buffer.getUint8(), buffer),
           extras: readValueOfType(buffer.getUint8(), buffer)
               ?.cast<String, String>(),
+          mediationExtras: readValueOfType(buffer.getUint8(), buffer)
+              ?.cast<List<MediationExtras>>(),
         );
+      case _valueMediationExtras:
+        // Returns null since there's no need to cast into MediationExtras
+        // instance
+        return null;
```
Comment (jjliu15) [inline packages/google_mobile_ads/lib/src/ad_instance_manager.dart:1074]: I generally prefer to just implement these so we can write unit tests. If we can cover this via android medium/ios xctests that's okay as well

**Review by jjliu15 — COMMENTED** [review]
Could you update the changelog? And the PR description as well

File: packages/google_mobile_ads/lib/src/mediation_extras.dart [hunk 6]
```diff
@@ -0,0 +1,39 @@
+// Copyright 2021 Google LLC
```
[the remark this answers is not in the document]
Comment (LTPhantom) [reply packages/google_mobile_ads/lib/src/mediation_extras.dart:1]: Good catch

**Comment by Phlphat746** [comment]
Thank you for you can working every once now!

**Review by luo86 — APPROVED** [review]

File: packages/google_mobile_ads/lib/src/ad_containers.dart [hunk 7]
```diff
@@ -233,6 +235,9 @@ class AdRequest {
   /// Extras to pass to the AdMob adapter.
   final Map<String, String>? extras;
 
+  /// Extra parameters to pass to specific ad adapters.
+  final List<MediationExtras>? mediationExtras;
```
Comment (jjliu15) [inline packages/google_mobile_ads/lib/src/ad_containers.dart:239]: Can you mark the `mediationExtrasIdentifier` field as deprecated?

File: packages/google_mobile_ads/ios/Classes/FLTGoogleMobileAdsPlugin.h [hunk 8]
```diff
@@ -67,14 +67,14 @@
                 mediationNetworkExtrasProvider
                                       registry:
                                           (id<FlutterPluginRegistry> _Nonnull)
-                                              registry;
+                                              registry __deprecated;
```
Comment (jjliu15) [inline packages/google_mobile_ads/ios/Classes/FLTGoogleMobileAdsPlugin.h:70]: Can you add a comment indicating to use MediationExtras instead?

File: packages/google_mobile_ads/ios/Classes/FLTMediationNetworkExtrasProvider.h [hunk 9]
```diff
@@ -19,6 +19,7 @@
  * An implementation of this protocol can be passed to FLTGoogleMobileAdsPlugin
  * using registerMediationNetworkExtrasProvider
  */
+__attribute__((deprecated))
```
Comment (jjliu15) [inline packages/google_mobile_ads/ios/Classes/FLTMediationNetworkExtrasProvider.h:22]: Can you update the comment to indicate what should be used instead?

**Review by jjliu15 — APPROVED** [review]

**Pull request merged**